### PR TITLE
Issue with oversell of products

### PR DESCRIPTION
**Pull request by o-iegorov**
New logic are introduced with reservation

### Description (*)
New reservation will be created before order placed (previous realization based on reservation after order placed). In case of errors within order placement additional reservation will be added to compensate

### Fixed Issues (if relevant)
https://jira.corp.magento.com/browse/MC-33900

### Manual testing scenarios (*)
When there is 1 quantity in stock and 2 orders placed to that product simultaneously both orders are going through and the saleable qty is updating to -1. Need to run simultaneously from 2 different browsers (or machines)

01. Add product to the cart which has 1 quantity left
02. Two customers placing orders at the same time for this SKU
03. Both orders will get success

EXPECTED RESULT
Only one order should go through

ACTUAL RESULT
Both orders going through and sellable quantity of the product updated to -1

Also please check functionality of the inventory:reservation:list-inconsistencies cli command with the next scenario: There are some reservations already exists, then update your msi with this changes, make some orders, check that command provide correct output. Place order with payment failure, check that command provide correct output

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by o-iegorov** [comment]
@magento run all tests

**Comment by o-iegorov** [comment]
@magento run all tests

**Comment by o-iegorov** [comment]
@magento run all tests

**Review by sidolov — CHANGES_REQUESTED** [review]
It would be great to add integration test for the negative cases when place order is failed.

**Comment by o-iegorov** [comment]
@magento run all tests

File: InventorySales/Test/Integration/Order/PlaceOrderWithException.php [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+use Magento\Catalog\Api\ProductRepositoryInterface;
+use Magento\Quote\Api\Data\CartItemInterface;
+use Magento\Quote\Api\Data\CartItemInterfaceFactory;
+use Magento\Quote\Api\CartRepositoryInterface;
+use Magento\Quote\Api\CartManagementInterface;
+use Magento\InventoryReservationsApi\Model\GetReservationsQuantityInterface;
+use Magento\TestFramework\Helper\Bootstrap;
+use PHPUnit\Framework\TestCase;
+
+class PlaceOrderWithException extends TestCase
```
Comment (sidolov) [inline InventorySales/Test/Integration/Order/PlaceOrderWithException.php]: Please, add *Test suffix to the class name

File: InventorySales/Test/Integration/Order/PlaceOrderWithException.php [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        $this->searchCriteriaBuilder = Bootstrap::getObjectManager()->get(SearchCriteriaBuilder::class);
+        $this->cartRepository = Bootstrap::getObjectManager()->get(CartRepositoryInterface::class);
+        $this->cartItemFactory = Bootstrap::getObjectManager()->get(CartItemInterfaceFactory::class);
+        $this->cartManagement = Bootstrap::getObjectManager()->get(CartManagementInterface::class);
+        $this->getReservationsQuantity = Bootstrap::getObjectManager()->get(GetReservationsQuantityInterface::class);
+    }
+
+    /**
+     * @magentoDataFixture Magento/Sales/_files/order.php
+     * @magentoDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/products.php
```
Comment (sidolov) [inline InventorySales/Test/Integration/Order/PlaceOrderWithException.php]: Please, use `@magentoDataFixture Magento_InventoryApi::Test/_files/products.php` instead for fixture path.

**Review by sidolov — CHANGES_REQUESTED** [review]

**Comment by sidolov** [comment]
@magento run all tests

**Comment by o-iegorov** [comment]
@magento run all tests

File: InventorySales/Test/Integration/Order/PlaceOrderWithException.php (hunk 1 above)
Comment (o-iegorov) [reply InventorySales/Test/Integration/Order/PlaceOrderWithException.php]: Added

File: InventorySales/Test/Integration/Order/PlaceOrderWithException.php (hunk 2 above)
Comment (o-iegorov) [reply InventorySales/Test/Integration/Order/PlaceOrderWithException.php]: Replaced

**Comment by sidolov** [comment]
@magento run all tests

**Comment by o-iegorov** [comment]
@magento run all tests

**Comment by sidolov** [comment]
@magento run all tests

File: InventorySales/Test/Integration/Order/PlaceOrderWithExceptionTest.php [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        $orderId = $this->cartManagement->placeOrder($cartId);
+        $salableQtyBefore = $this->getReservationsQuantity->execute($sku, $stockId);
+
+        self::expectException(\Exception::class);
+        $this->cartManagement->placeOrder($cartId);
+
+        $salableQtyAfter = $this->getReservationsQuantity->execute($sku, $stockId);
+        self::assertSame($salableQtyBefore, $salableQtyAfter);
+
+        $this->deleteOrderById((int)$orderId);
```
Comment (sidolov) [inline InventorySales/Test/Integration/Order/PlaceOrderWithExceptionTest.php]: Method `deleteOrderById` does not exist

**Review by sidolov — CHANGES_REQUESTED** [review]

File: InventorySales/Test/Integration/Order/PlaceOrderWithExceptionTest.php (hunk 3 above)
Comment (o-iegorov) [reply InventorySales/Test/Integration/Order/PlaceOrderWithExceptionTest.php]: Fixed

**Comment by o-iegorov** [comment]
@magento run all tests

**Comment by o-iegorov** [comment]
@magento run all tests

**Comment by o-iegorov** [comment]
@magento run Integration Tests

**Comment by o-iegorov** [comment]
@magento run Integration Tests

**Comment by o-iegorov** [comment]
@magento run Integration Tests

**Comment by o-iegorov** [comment]
@magento run all tests

**Comment by o-iegorov** [comment]
@magento run Integration Tests

**Comment by o-iegorov** [comment]
@magento run Integration Tests

**Comment by o-iegorov** [comment]
@magento run Integration Tests

**Comment by o-iegorov** [comment]
@magento run Integration Tests

**Comment by o-iegorov** [comment]
@magento run Integration Tests

**Comment by o-iegorov** [comment]
@magento run all tests

**Comment by sidolov** [comment]
SVC approval ticket: https://jira.corp.magento.com/browse/MC-35247

**Comment by sidolov** [comment]
@magento run all tests
